pkg/agent/cniserver/interface_configuration_linux.go

```diff
@@ -92,49 +93,170 @@ func (ic *ifConfigurator) moveIfToNetns(ifname string, netns ns.NetNS) error {
+		tempNSDev, err := ic.netlink.LinkByName(vfNetDevice)
 		if err != nil {
-			return fmt.Errorf("failed to set MTU for VF netdevice %s: %v", containerIfaceName, err)
+			return fmt.Errorf("failed to find %s in tempNS: %w", vfNetDevice, err)
+		}
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				tmpErr := ic.netlink.LinkSetNsFd(tempNSDev, int(hostNS.Fd()))
+				if tmpErr != nil {
+					klog.ErrorS(tmpErr, "Failed to set network namespace to hostNS for the device", "device", tempNSDev.Attrs().Name, "hostNS", hostNS.Path())
+				}
+			}
+		}()
+
+		// Rename the device to the wanted name
+		if err = ic.netlink.LinkSetName(tempNSDev, containerIfaceName); err != nil {
+			return fmt.Errorf("failed to rename VF device %s to %s: %w", vfNetDevice, containerIfaceName, err)
 		}
-		err = ic.netlink.LinkSetUp(link)
+
+		// Restore the original device name in case of error
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				tmpErr := ic.netlink.LinkSetName(tempNSDev, vfNetDevice)
+				if tmpErr != nil {
+					klog.ErrorS(tmpErr, "Failed to set device name back", "device", tempNSDev.Attrs().Name, "originalName", vfNetDevice)
+				}
+			}
+		}()
+
+		err = ic.netlink.LinkSetAlias(tempNSDev, vfNetDevice)
 		if err != nil {
-			return fmt.Errorf("failed to set link up to VF netdevice %s: %v", containerIfaceName, err)
+			return fmt.Errorf("failed to set alias as %s for VF netdevice %s: %w", vfNetDevice, vfNetDevice, err)
 		}
-		containerIface.Name = containerIfaceName
-		containerIface.Mac = link.Attrs().HardwareAddr.String()
-		containerIface.Sandbox = netns.Path()
-		klog.V(2).Infof("hostIface: %+v, containerIface: %+v", hostIface, containerIface)
-		klog.V(2).Infof("Configuring IP address for container %s", containerID)
-		// result.Interfaces must be set before this.
-		if err := ipamConfigureIface(containerIface.Name, result); err != nil {
-			return fmt.Errorf("failed to configure IP address for container %s: %v", containerID, err)
+		klog.V(2).InfoS("Link's alias has been updated", "alias", vfNetDevice)
+
+		// Remove the alias on error
+		defer func() {
+			if err != nil && tempNSDev != nil {
+				tmpErr := ic.netlink.LinkSetAlias(tempNSDev, "")
+				if tmpErr != nil {
+					klog.ErrorS(tmpErr, "Failed to empty the alias", "device", tempNSDev.Attrs().Name)
+				}
+			}
+		}()
+
+		// Move VF to Container namespace
+		err = ic.moveIFToNetNS(containerIfaceName, netns)
+		if err != nil {
+			return fmt.Errorf("failed to move VF %s to container netns %s: %w", vfNetDevice, containerNetNS, err)
 		}
-		klog.V(2).Infof("ipam.ConfigureIface result: %+v, err: %v", result, err)
-		return nil
+
+		// Look up the device again on error, as the index might have changed
+		defer func() {
+			if err != nil {
+				var tmpErr error
+				klog.InfoS("Look up the device again on error, as the index might have changed")
+				tempNSDev, tmpErr = ic.netlink.LinkByName(containerIfaceName)
```

Comment:
You did not reply to my earlier question:

> I meant when the interface is not in the root NS, could ic.netlink.LinkByName(containerIfaceName) still find the interface by name?

Reply:
yeah, I believe so, we can leverage on Pod deletion to recover VF in that case.

Reply:
Ok, could we add a comment here, saying if the link lookup fails here, remaining cleanup will not be executed?